### PR TITLE
Fix parser to continue parsing key-value pairs after an If-Statement value in a HashExpression

### LINKED ISSUE #6970: Hashtable literal parsing fails when using an if statement as value
If there are other elements after the `if() {}` part, then parser will fail.
But it's fine if you have `if() {} else {}` as the value.

Steps to reproduce
------------------

```powershell
$script = @'
@{
 a = if($true) { 'true' }
 b = 10
}
'@
$er = $null
$a = [System.Management.Automation.Language.Parser]::ParseInput($script, [ref]$null, [ref]$er)
$er
```

Expected behavior
-----------------

`$er` is null. No error produced.

Actual behavior
---------------

```none
Extent ErrorId              Message                                                    IncompleteInput
------ -------              -------                                                    ---------------
       MissingEndCurlyBrace Missing closing '}' in statement block or type definition.           False
b      UnexpectedToken      Unexpected token 'b' in expression or statement.                     False
}      UnexpectedToken      Unexpected token '}' in expression or statement.                     False
```

Environment data
----------------

```powershell
> $PSVersionTable
Name                           Value
----                           -----
PSVersion                      6.0.1
PSEdition                      Core
GitCommitId                    v6.0.1
OS                             Microsoft Windows 10.0.17134
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by daxian-dbw**
## PR Summary

Fix #6970
After parsing `if () { }`, new lines are skipped in order to see if either `elseif` or `else` is present. When neither is present, we should resync back to the pointer before skipping those possibly available new lines, so that the new line tokens can be utilized by the subsequent parsing.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [x] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [x] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

File: src/System.Management.Automation/engine/parser/Parser.cs [hunk 1]
```diff
@@ -2411,6 +2411,16 @@ private StatementAst IfStatementRule(Token ifToken)
...
+                // Save a restore point here. In case there is no 'elseif' or 'else' following,
+                // we should resync back here to preserve the possible new lines. The new lines
+                // could be important for the following parsing. For example, in case we are in
+                // a HashExpression, a new line might be needed for parsing the key-value that
+                // is following the if statement:
+                //    @{
+                //       a = if (1) {}
+                //       b = 10
+                //    }
+                var restorePoint = PeekToken();
```
[the remark this answers is not in the document]
Comment (daxian-dbw) [reply src/System.Management.Automation/engine/parser/Parser.cs]: I decide to directly check `_ungotToken == null`. I think this way will deliver the intent more clearly.

File: test/powershell/Language/Parser/LanguageAndParser.TestFollowup.Tests.ps1 [hunk 2]
```diff
@@ -212,3 +212,22 @@ Describe "Members of System.Type" -Tags "CI" {
         [MyType].ImplementedInterfaces | Should -Be System.Collections.IEnumerable
     }
 }
+
+Describe "Hash expression with if statement as value" -Tags "CI" {
+    It "Key-value pairs after an if-statement-value in a HashExpression should continue to be parsed" {
+        $result = @{
+            a = if (1) {'a'}
+            b = 'b'
+            c = if (0) {2} elseif (1) {'c'}
```
Comment (lzybkr) [inline test/powershell/Language/Parser/LanguageAndParser.TestFollowup.Tests.ps1]: Maybe also test:
* newlines after `}`
* if/elseif/else

**Review by lzybkr — APPROVED** [review]

File: test/powershell/Language/Parser/LanguageAndParser.TestFollowup.Tests.ps1 (hunk 2 above)
Comment (daxian-dbw) [reply test/powershell/Language/Parser/LanguageAndParser.TestFollowup.Tests.ps1]: Fixed. Thanks!

**Comment by daxian-dbw** [comment]
@rjmholt I guess you see that use from https://docs.microsoft.com/en-us/powershell/gallery/concepts/module-psedition-support#sample-module-manifest-file-with-compatiblepseditions-key :) The current implementation supports that.

**Comment by rjmholt** [comment]
Yes you're quite right. I was experimenting with it before and must have done something wrong.

**Comment by rjmholt** [comment]
Might be worth including a test with that scenario

**Review by rjmholt — APPROVED** [review]

File: test/powershell/Language/Parser/LanguageAndParser.TestFollowup.Tests.ps1 [hunk 3]
```diff
@@ -212,3 +212,94 @@ Describe "Members of System.Type" -Tags "CI" {
...
+            d = 'd'
+            e = if (0) {2} elseif (0) {2} else {'e'}
+            f = 'f'
+            g = if (0) {2} else {'g'}
+            h = 'h'
+        }
+
+        # With extra new lines after if-statement
```
Comment (iSazonov) [inline test/powershell/Language/Parser/LanguageAndParser.TestFollowup.Tests.ps1]: Perhaps the comment should be more stringent
`Caution!!! Don't remove extra new lines after if-statement!`

**Pull request merged**